### PR TITLE
Adding functionality for increasing host_to_device_transfer_threads

### PR DESCRIPTION
**Pull request by chandrasekhard2**
Increasing the host to device transfer threads to 4 improved the ResNet Performance by 20 seconds on v4-8.

### CODE REVIEW

File: torch_xla/distributed/parallel_loader.py [hunk 1]
```diff
@@ -95,9 +99,10 @@ def __init__(self,
     thread.daemon = True
     thread.start()
     for dqueue in itervalues(self._queues):
-      thread = threading.Thread(target=self._worker, args=(dqueue,))
-      thread.daemon = True
-      thread.start()
+      for i in range(host_to_device_transfer_threads):
+        thread = threading.Thread(target=self._worker, args=(dqueue,))
+        thread.daemon = True
+        thread.start()
```
Comment (JackCaoG) [inline torch_xla/distributed/parallel_loader.py:105]: Can we also modify the resnet example to use this `host_to_device_transfer_threads`? or is that going to cause the accuracy issue compared to the origional implementation?

Comment (chandrasekhard2) [reply torch_xla/distributed/parallel_loader.py:105]: It will only improve the performance and doesn't affect the convergence.

**Comment by chandrasekhard2** [comment]
Output with fake data

==> Preparing data..
==> Preparing data..
==> Preparing data..
==> Preparing data..
Epoch 1 train begin 01:17:05
| Training Device=xla:0/1 Epoch=1 Step=0 Loss=6.89059 Rate=1.79 GlobalRate=1.79 Time=01:18:15
| Training Device=xla:0/2 Epoch=1 Step=0 Loss=6.89059 Rate=1.71 GlobalRate=1.71 Time=01:18:18
| Training Device=xla:0/3 Epoch=1 Step=0 Loss=6.89059 Rate=1.69 GlobalRate=1.69 Time=01:18:19
| Training Device=xla:0/0 Epoch=1 Step=0 Loss=6.89059 Rate=1.72 GlobalRate=1.72 Time=01:18:19
| Training Device=xla:0/0 Epoch=1 Step=300 Loss=0.00902 Rate=313.35 GlobalRate=260.43 Time=01:19:33
| Training Device=xla:0/3 Epoch=1 Step=300 Loss=0.00902 Rate=310.97 GlobalRate=256.63 Time=01:19:33
| Training Device=xla:0/1 Epoch=1 Step=300 Loss=0.00902 Rate=298.07 GlobalRate=258.26 Time=01:19:33
| Training Device=xla:0/2 Epoch=1 Step=300 Loss=0.00902 Rate=306.60 GlobalRate=256.83 Time=01:19:33
| Training Device=xla:0/3 Epoch=1 Step=600 Loss=0.00287 Rate=1217.39 GlobalRate=449.32 Time=01:19:54
| Training Device=xla:0/1 Epoch=1 Step=600 Loss=0.00287 Rate=1212.23 GlobalRate=451.82 Time=01:19:54
| Training Device=xla:0/2 Epoch=1 Step=600 Loss=0.00287 Rate=1215.65 GlobalRate=449.62 Time=01:19:54
| Training Device=xla:0/0 Epoch=1 Step=600 Loss=0.00287 Rate=1218.33 GlobalRate=455.15 Time=01:19:54
| Training Device=xla:0/2 Epoch=1 Step=900 Loss=0.00178 Rate=1578.71 GlobalRate=600.09 Time=01:20:15
| Training Device=xla:0/0 Epoch=1 Step=900 Loss=0.00178 Rate=1579.79 GlobalRate=606.65 Time=01:20:15
| Training Device=xla:0/3 Epoch=1 Step=900 Loss=0.00178 Rate=1579.40 GlobalRate=599.73 Time=01:20:15
| Training Device=xla:0/1 Epoch=1 Step=900 Loss=0.00178 Rate=1577.33 GlobalRate=602.70 Time=01:20:15
| Training Device=xla:0/3 Epoch=1 Step=1200 Loss=0.00146 Rate=1725.50 GlobalRate=720.49 Time=01:20:36
| Training Device=xla:0/2 Epoch=1 Step=1200 Loss=0.00146 Rate=1725.21 GlobalRate=720.88 Time=01:20:36
| Training Device=xla:0/1 Epoch=1 Step=1200 Loss=0.00146 Rate=1724.67 GlobalRate=723.70 Time=01:20:36
| Training Device=xla:0/0 Epoch=1 Step=1200 Loss=0.00146 Rate=1725.65 GlobalRate=727.97 Time=01:20:36
| Training Device=xla:0/3 Epoch=1 Step=1500 Loss=0.00138 Rate=1783.59 GlobalRate=819.52 Time=01:20:57
| Training Device=xla:0/2 Epoch=1 Step=1500 Loss=0.00138 Rate=1783.48 GlobalRate=819.93 Time=01:20:57
| Training Device=xla:0/0 Epoch=1 Step=1500 Loss=0.00138 Rate=1783.64 GlobalRate=827.26 Time=01:20:57
| Training Device=xla:0/1 Epoch=1 Step=1500 Loss=0.00138 Rate=1783.25 GlobalRate=822.85 Time=01:20:57
| Training Device=xla:0/3 Epoch=1 Step=1800 Loss=0.00135 Rate=1806.21 GlobalRate=902.18 Time=01:21:18
| Training Device=xla:0/1 Epoch=1 Step=1800 Loss=0.00135 Rate=1806.09 GlobalRate=905.54 Time=01:21:18
| Training Device=xla:0/0 Epoch=1 Step=1800 Loss=0.00135 Rate=1806.25 GlobalRate=909.99 Time=01:21:18
| Training Device=xla:0/2 Epoch=1 Step=1800 Loss=0.00135 Rate=1806.16 GlobalRate=902.59 Time=01:21:18
| Training Device=xla:0/3 Epoch=1 Step=2100 Loss=0.00135 Rate=1816.07 GlobalRate=972.30 Time=01:21:39
| Training Device=xla:0/2 Epoch=1 Step=2100 Loss=0.00135 Rate=1816.06 GlobalRate=972.71 Time=01:21:39
| Training Device=xla:0/1 Epoch=1 Step=2100 Loss=0.00135 Rate=1816.02 GlobalRate=975.64 Time=01:21:39
| Training Device=xla:0/0 Epoch=1 Step=2100 Loss=0.00135 Rate=1816.08 GlobalRate=980.07 Time=01:21:39
Epoch 1 train end 01:22:23
| Test Device=xla:0/3 Step=0 Epoch=1 Time=01:22:27
| Test Device=xla:0/1 Step=0 Epoch=1 Time=01:22:29
| Test Device=xla:0/0 Step=0 Epoch=1 Time=01:22:30
| Test Device=xla:0/2 Step=0 Epoch=1 Time=01:22:33
Epoch 1 test end 01:22:43, Accuracy=100.00
Max Accuracy: 100.00%

File: test/test_train_mp_imagenet.py [hunk 2]
```diff
@@ -76,11 +91,17 @@
...
 DEFAULT_KWARGS = dict(
     batch_size=128,
-    test_set_batch_size=64,
+    test_set_batch_size=128,
     num_epochs=18,
     momentum=0.9,
     lr=0.1,
     target_accuracy=0.0,
+    persistent_workers=True,
+    prefetch_factor=32,
```
Comment (miladm) [inline test/test_train_mp_imagenet.py:100]: You should probably note that these configurations are enable best perf results on, say, v4-8. This way users know that for other hardware configs, they probably need to make adjustments.

File: torch_xla/distributed/parallel_loader.py [hunk 3]
```diff
@@ -81,7 +84,8 @@ def __init__(self,
                batchdim=0,
                batches_per_execution=1,
                loader_prefetch_size=8,
-               device_prefetch_size=4):
+               device_prefetch_size=4,
```
Comment (miladm) [inline torch_xla/distributed/parallel_loader.py:87]: why did you dial this down?

File: torch_xla/distributed/parallel_loader.py [hunk 4]
```diff
@@ -81,7 +84,8 @@ def __init__(self,
...
+               host_to_device_transfer_threads=1):
```
Comment (miladm) [inline torch_xla/distributed/parallel_loader.py:88]: ditto

**Review by miladm — COMMENTED** [review]
Thanks @chandrasekhard2. Left a few comments here.

File: test/test_train_mp_imagenet.py (hunk 2 above)
Comment (miladm) [reply test/test_train_mp_imagenet.py:100]: I'd be in favor of creating a map data structure that holds all keys needed to get best perf results. Then, pass the map value to the appropriate fields throughout the script.

**Comment by JackCaoG** [comment]
I think the issue is you merge the master into your branch, instead of doing the rebase. Our CI doesn't like that lol. One quick fix is to take the diff and open the new pr, so you can play with the commits.

**Comment by miladm** [comment]
+1

@chandrasekhard2 can you please prioritize this change to land?
